docs/design/backups.md

```diff
@@ -0,0 +1,83 @@
+# Backup & Restore
+
+In particularly disastrous situations, we would like to be able to revert the etcd cluster to a previously known healthy state.
+This document lays out the design of how we want the healthy state to be captured, securely stored & safely restored to operator-managed etcd clusters.
+
+## Backup Resource
+
+We will introduce a new custom resource: `EtcdBackup`.
+
+```yaml
+apiVersion: etcd.improbable.io/v1alpha1
+kind: EtcdBackup
+metadata:
+  name: my-cluster-backup
+  namespace: default
+spec:
+  type: snapshot
+  destination:
+    gcsBucket:
+      bucketName: my-cluster-backup-bucket
+      creds:
+        secretRef: 
+          name: my-cluster-backup-bucket-credentials
+          namespace: default
+```
+
+This resource specifies that we desire a backup to be immediately taken as a [snapshot](https://github.com/etcd-io/etcd/blob/master/Documentation/op-guide/recovery.md#snapshotting-the-keyspace) from the etcd API.
+This snapshot will be saved to a GCS bucket, accessed with credentials read from a given secret.
+The file would be named with a timestamp when the backup was started, for uniqueness.
+Backups will be taken by the controller, rather than a Job or Pod acting on behalf of the controller.
+
+Backup types could include
+  * `snapshot`: a `snapshot.db` file taken from the etcd API
+  * `keys`: a zip file containing all key-values as files
+  * `keys-json` a json file containing all key-values as a JSON file
```

Comment:
Are these last 2 standard formats? What's the purpose here? Audit?

Reply:
I thought the `snapshot.db` file already contained everything. Why duplicate them as `keys` and `keys-json`?

Reply:
i mostly put these here to demonstrate that there are possibly alternatives.

it's possibly a less destructive restore story to write a bunch of keys into an existing cluster, rather than dealing with snapshot restores. at Improbable we currently dump all keys as a tarball for precisely that reason - our infrastructure doesn't handle restoring from a snapshot well. 

but as said, we'll only implement the snapshot type for now as that will be fine for most users.